{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "7f67bf21-32da-4a56-ab8b-eb9725aa0705",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/usr/bin/env python3\n",
    "#\n",
    "# FileSender www.filesender.org\n",
    "#\n",
    "# Copyright (c) 2009-2019, AARNet, Belnet, HEAnet, SURFnet, UNINETT\n",
    "# All rights reserved.\n",
    "#\n",
    "# Redistribution and use in source and binary forms, with or without\n",
    "# modification, are permitted provided that the following conditions are met:\n",
    "#\n",
    "# *   Redistributions of source code must retain the above copyright\n",
    "#     notice, this list of conditions and the following disclaimer.\n",
    "# *   Redistributions in binary form must reproduce the above copyright\n",
    "#     notice, this list of conditions and the following disclaimer in the\n",
    "#     documentation and/or other materials provided with the distribution.\n",
    "# *   Neither the name of AARNet, Belnet, HEAnet, SURFnet and UNINETT nor the\n",
    "#     names of its contributors may be used to endorse or promote products\n",
    "#     derived from this software without specific prior written permission.\n",
    "#\n",
    "# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS \"AS IS\"\n",
    "# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE\n",
    "# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE\n",
    "# DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE\n",
    "# FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL\n",
    "# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR\n",
    "# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER\n",
    "# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,\n",
    "# OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE\n",
    "# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "66078e65-c0b5-4276-a76e-e1644b6437cd",
   "metadata": {},
   "outputs": [],
   "source": [
    "import argparse\n",
    "import requests\n",
    "import time\n",
    "from collections.abc import Iterable\n",
    "from collections.abc import MutableMapping\n",
    "import hmac\n",
    "import hashlib\n",
    "import urllib3\n",
    "import os\n",
    "import json\n",
    "import configparser\n",
    "from os.path import expanduser\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "e43062d1-57a0-4228-b6e9-fb8290c7a219",
   "metadata": {},
   "outputs": [],
   "source": [
    "from multiprocessing import Pool\n",
    "from functools import partial\n",
    "import time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d2d8438e-ad3b-4b59-b4eb-2291e9092a9b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e201835c-1cb5-4f08-8918-605cd830e6ab",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "7e42ff3d-9a5d-42c8-b2c9-142d40835252",
   "metadata": {},
   "outputs": [],
   "source": [
    "urllib3.disable_warnings(urllib3.exceptions.InsecureRequestWarning)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "654a77b4-019c-489c-9fde-1a794ede5091",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "480d968a-9f6c-4f2f-91f5-8e271dee21b6",
   "metadata": {},
   "outputs": [],
   "source": [
    "def flatten(d, parent_key=''):\n",
    "    items = []\n",
    "    for k, v in d.items():\n",
    "        new_key = parent_key + '[' + k + ']' if parent_key else k\n",
    "        if isinstance(v, MutableMapping):\n",
    "            items.extend(flatten(v, new_key).items())\n",
    "        else:\n",
    "            items.append(new_key+'='+v)\n",
    "    items.sort()\n",
    "    return items\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "5c379d35-eac7-4b18-a02f-8c006e985ada",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def call(method, path, data, content=None, rawContent=None, options={}):\n",
    "    data['remote_user'] = username\n",
    "    data['timestamp'] = str(round(time.time()))\n",
    "    flatdata = flatten(data)\n",
    "    signed = bytes(method+'&'+base_url.replace('https://', '',\n",
    "                   1).replace('http://', '', 1)+path+'?'+('&'.join(flatten(data))), 'ascii')\n",
    "\n",
    "    content_type = options['Content-Type'] if 'Content-Type' in options else 'application/json'\n",
    "\n",
    "    inputcontent = None\n",
    "    if content is not None and content_type == 'application/json':\n",
    "        inputcontent = json.dumps(content, separators=(',', ':'))\n",
    "        signed += bytes('&'+inputcontent, 'ascii')\n",
    "    elif rawContent is not None:\n",
    "        inputcontent = rawContent\n",
    "        signed += bytes('&', 'ascii')\n",
    "        signed += inputcontent\n",
    "\n",
    "    # print(signed)\n",
    "    bkey = bytearray()\n",
    "    bkey.extend(map(ord, apikey))\n",
    "    data['signature'] = hmac.new(bkey, signed, hashlib.sha1).hexdigest()\n",
    "\n",
    "    url = base_url+path+'?'+('&'.join(flatten(data)))\n",
    "    headers = {\n",
    "        \"Accept\": \"application/json\",\n",
    "        \"Content-Type\": content_type\n",
    "    }\n",
    "    response = None\n",
    "    if method == \"get\":\n",
    "        response = requests.get(url, verify=not insecure, headers=headers)\n",
    "    elif method == \"post\":\n",
    "        response = requests.post(\n",
    "            url, data=inputcontent, verify=not insecure, headers=headers)\n",
    "    elif method == \"put\":\n",
    "        response = requests.put(url, data=inputcontent,\n",
    "                                verify=not insecure, headers=headers)\n",
    "    elif method == \"delete\":\n",
    "        response = requests.delete(url, verify=not insecure, headers=headers)\n",
    "\n",
    "    if response is None:\n",
    "        raise Exception('Client error')\n",
    "\n",
    "    code = response.status_code\n",
    "    # print(url)\n",
    "    # print(inputcontent)\n",
    "    # print(code)\n",
    "    # print(response.text)\n",
    "\n",
    "    if code != 200:\n",
    "        if method != 'post' or code != 201:\n",
    "            raise Exception('Http error '+str(code)+' '+response.text)\n",
    "\n",
    "    if response.text == \"\":\n",
    "        raise Exception('Http error '+str(code)+' Empty response')\n",
    "\n",
    "    if method != 'post':\n",
    "        return response.json()\n",
    "\n",
    "    r = {}\n",
    "    r['location'] = response.headers['Location']\n",
    "    r['created'] = response.json()\n",
    "    return r\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "418046c8-2a6c-4f9e-b256-2876d05000b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def postTransfer(user_id, files, recipients, subject=None, message=None, expires=None, options=[]):\n",
    "\n",
    "    if expires is None:\n",
    "        expires = round(time.time()) + (default_transfer_days_valid*24*3600)\n",
    "\n",
    "    to = [x.strip() for x in recipients.split(',')]\n",
    "    print(to)\n",
    "\n",
    "    return call(\n",
    "        'post',\n",
    "        '/transfer',\n",
    "        {},\n",
    "        {\n",
    "            'from': user_id,\n",
    "            'files': files,\n",
    "            'recipients': to,\n",
    "            'subject': subject,\n",
    "            'message': message,\n",
    "            'expires': expires,\n",
    "            'aup_checked': 1,\n",
    "            'options': options\n",
    "        },\n",
    "        None,\n",
    "        {}\n",
    "    )\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "52b11dd0-be7b-4b4c-b241-a266a0ab3f41",
   "metadata": {},
   "outputs": [],
   "source": [
    "def putChunk(t, f, chunk, offset):\n",
    "    return call(\n",
    "        'put',\n",
    "        '/file/'+str(f['id'])+'/chunk/'+str(offset),\n",
    "        {'key': f['uid'], 'roundtriptoken': t['roundtriptoken']},\n",
    "        None,\n",
    "        chunk,\n",
    "        {'Content-Type': 'application/octet-stream'}\n",
    "    )\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "6732e809-cdc7-40de-bc0a-ac0b1b99dced",
   "metadata": {},
   "outputs": [],
   "source": [
    "def fileComplete(t,f):\n",
    "    return call(\n",
    "        'put',\n",
    "        '/file/'+str(f['id']),\n",
    "        { 'key': f['uid'], 'roundtriptoken': t['roundtriptoken'] },\n",
    "        { 'complete': True },\n",
    "        None,\n",
    "        {}\n",
    "    )\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "3b4352da-9972-4597-82c1-1c25e6978167",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "def deleteTransfer(transfer):\n",
    "    return call(\n",
    "        'delete',\n",
    "        '/transfer/'+str(transfer['id']),\n",
    "        {'key': transfer['files'][0]['uid']},\n",
    "        None,\n",
    "        None,\n",
    "        {}\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "6b4fff1c-94de-475d-9222-eafae78400bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def transferComplete(transfer):\n",
    "    return call(\n",
    "        'put',\n",
    "        '/transfer/'+str(transfer['id']),\n",
    "        {'key': transfer['files'][0]['uid']},\n",
    "        {'complete': True},\n",
    "        None,\n",
    "        {}\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "a10734ce-814b-4e1e-9bfa-eac4063ce5b8",
   "metadata": {},
   "outputs": [],
   "source": [
    "def upload_file( fileobject, transferData, filesData, upload_chunk_size, debug):\n",
    "    fname = fileobject[\"name\"]\n",
    "    fsize = fileobject[\"size\"]\n",
    "    fstring = f\"{fname}:{fsize}\"\n",
    "    fpath = filesData[fstring][\"path\"]\n",
    "\n",
    "    try:\n",
    "        # putChunks\n",
    "        if debug:\n",
    "            print('putChunks: '+fpath)\n",
    "        with open(fpath, mode='rb', buffering=0) as fin:\n",
    "            chunk_count = 0\n",
    "            for offset in range(0, fsize, upload_chunk_size):\n",
    "                if progress:\n",
    "                    print('Uploading: '+fpath+' '+str(offset)+'-'+str(min(offset +\n",
    "                          upload_chunk_size, fsize))+' '+str(round(offset/fsize*100))+'%')\n",
    "                data = fin.read(upload_chunk_size)\n",
    "                # print(data)\n",
    "                putChunk(transferData, fileobject, data, offset)\n",
    "                if debug:\n",
    "                    chunk_count += 1\n",
    "                    print(f\"uploaded {chunk_count} chunks\")\n",
    "        # file complete\n",
    "        if debug:\n",
    "            print('fileComplete: '+fpath)\n",
    "        res = fileComplete(transferData, fileobject)\n",
    "        if progress:\n",
    "            print('Uploading: '+fpath+' '+str(size)+' 100%')\n",
    "        return res\n",
    "    except Exception as e:\n",
    "        raise(e)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "53cf5b3a-b84a-4ab8-8f93-d51ae5db429f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def transfer_data_to_text(tdata):\n",
    "    total_size = 0\n",
    "    for f in tdata[\"files\"]:\n",
    "        total_size += f[\"size\"]\n",
    "    size_unit = \"B\"\n",
    "    if total_size > 1024**3:\n",
    "        size_unit = \"GB\"\n",
    "        total_size = total_size/(1024**3)\n",
    "    elif total_size > 1024**2:\n",
    "        size_unit = \"MB\"\n",
    "        total_size = total_size/(1024**2)\n",
    "    elif total_size > 1024:\n",
    "        size_unit = \"kB\"\n",
    "        total_size = total_size/(1024)\n",
    "    size_str = f\"{total_size:.2f} {size_unit}\"\n",
    "    url = res[\"recipients\"][0][\"download_url\"]\n",
    "    report_txt = f\"\"\"FileSender upload\n",
    "transfer ID: {tdata[\"id\"]}\n",
    "uploaded by: {tdata[\"user_email\"]}\n",
    "total size:  {size_str}\n",
    "\n",
    "recipient:   {tdata[\"recipients\"][0][\"email\"]}\n",
    "D/L link:    {url}\n",
    "\n",
    "Files uploaded:\n",
    "\"\"\"\n",
    "\n",
    "    for fidx, f in enumerate(tdata[\"files\"]):\n",
    "        url = f\"https://filesender.aarnet.edu.au/download.php?token={res['recipients'][0]['token']}&files_ids={f['id']}\"\n",
    "        ct_str = f\"{fidx}\".rjust( len(str(len(tdata[\"files\"]))) )\n",
    "        report_txt += f\"\"\"\n",
    "{ct_str}: \"{f[\"name\"]}\"\n",
    "{\" \"*len(ct_str)}  {f[\"size\"]:,} bytes \n",
    "{\" \"*len(ct_str)}  {url}&files_ids={f[\"id\"]}\n",
    "\"\"\"\n",
    "\n",
    "    return report_txt\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "efe04d39-86f5-4373-955d-f30c0e4f6e21",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8998e34a-f65a-4ce5-9a42-f3f02fa59b66",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "49963e35-fa42-433c-bdaa-7adb5a6599a2",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a2a48ade-2245-48c3-9090-537dd87d4297",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "93cc2189-de36-4daf-8a02-9fdcb0be540a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "12ac66cf-169a-41a3-9faf-08fdad64a09b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4858c29f-b04e-4939-8c6f-5b590e9c1039",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8b9ead64-69a1-44dd-876a-6b1040b1e081",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4fd87ae1-9f8e-499f-a392-b88f50620972",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a368c49d-0312-425b-a58c-5f7540720492",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "e01afb5d-24fd-4f89-979e-bda841907979",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "#argv\n",
    "parser = argparse.ArgumentParser()\n",
    "parser.add_argument(\"files\", help=\"path to file(s) to send\", nargs='+')\n",
    "parser.add_argument(\"-v\", \"--verbose\", action=\"store_true\")\n",
    "parser.add_argument(\"-i\", \"--insecure\", action=\"store_true\")\n",
    "parser.add_argument(\"-p\", \"--progress\", action=\"store_true\")\n",
    "parser.add_argument(\"-s\", \"--subject\")\n",
    "parser.add_argument(\"-m\", \"--message\")\n",
    "parser.add_argument(\"-k\", \"--skip-email\", action=\"store_true\", default=False, help=\"Don't send email to recipient\")\n",
    "\n",
    "requiredNamed = parser.add_argument_group('required named arguments')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "40d9bcb2-b50b-49e6-8fdf-222130d14dbc",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f8dd617c-245b-48db-a096-6a3aba465b97",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "c9385af6-f947-4749-b8b2-5a66665ae225",
   "metadata": {},
   "outputs": [],
   "source": [
    "n_procs = 4"
   ]
  },
  {
   "cell_type": "raw",
   "id": "99a86f1c-3c0d-458e-b69d-89bfb768d7db",
   "metadata": {},
   "source": [
    "filelist = [\"/home/paul/Downloads/bowtie-1.3.1-linux-x86_64.zip\",\n",
    "            \"/home/paul/Zotero/storage/QS8AEKXV/Gong et al. - 2022 - Analysis and Performance Assessment of the Whole G.pdf\",\n",
    "            \"/home/paul/Downloads/I100_issuu_sub.pdf\"]\n",
    "n_procs = min(n_procs, len(filelist))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "d09c46bc-72c5-4a64-a30a-858fbc4067a1",
   "metadata": {},
   "outputs": [],
   "source": [
    "n_procs = 4\n",
    "\n",
    "filelist = \"\"\"/home/paul/Zotero/storage/28F8IG85/Lokody - 2013 - Cancer genomics Signature analysis suggests cance.pdf\n",
    "/home/paul/Zotero/storage/29G6QV82/Usuyama et al. - 2014 - HapMuC somatic mutation calling using heterozygou.pdf\n",
    "/home/paul/Zotero/storage/2BBJ65XW/Pleasance et al. - 2010 - A small cell lung cancer genome reports complex to.pdf\n",
    "/home/paul/Zotero/storage/2CR2WGUC/Taub et al. - 2010 - Overcoming bias and systematic errors in next gene.pdf\n",
    "/home/paul/Zotero/storage/2DRPFWV8/Kim et al. - 2016 - Somatic ERCC2 mutations are associated with a dist.pdf\n",
    "/home/paul/Zotero/storage/2FXQNZ3S/Alter et al. - 2000 - Singular value decomposition for genome-wide expre.pdf\n",
    "/home/paul/Zotero/storage/2GTA3FY6/Garcia-Prieto et al. - 2020 - The consequences of variant calling decisions in s.pdf\n",
    "/home/paul/Zotero/storage/2JRP4NTD/Kim et al. - 2018 - Strelka2 fast and accurate calling of germline an.pdf\n",
    "/home/paul/Zotero/storage/2P6RPSFQ/Pellegrino et al. - 2018 - High-throughput single-cell DNA sequencing of acut.pdf\n",
    "/home/paul/Zotero/storage/2QNXMWPI/Kandoth et al. - 2013 - Mutational landscape and significance across 12 ma.pdf\n",
    "/home/paul/Zotero/storage/2S6DPCZ7/Liu et al. - 2016 - Joint detection of copy number variations in paren.pdf\n",
    "/home/paul/Zotero/storage/2UA83RGN/Consortium - 2012 - An integrated map of genetic variation from 1,092 .pdf\n",
    "/home/paul/Zotero/storage/333VUI79/Zhu-SII09.pdf\n",
    "/home/paul/Zotero/storage/34C3ERMA/Bersanelli et al. - 2016 - Methods for the integration of multi-omics data m.pdf\n",
    "/home/paul/Zotero/storage/3BGPU5KB/Hansen et al. - 2012 - BSmooth from whole genome bisulfite sequencing re.pdf\n",
    "/home/paul/Zotero/storage/3CFDCRFR/O'Brien et al. - 2010 - Gremlin An Interactive Visualization Model for An.pdf\n",
    "/home/paul/Zotero/storage/3HBZH2XM/gim2014178a.pdf\n",
    "/home/paul/Zotero/storage/3HMKCBF3/Parisi et al. - 2014 - Ranking and combining multiple predictors without .pdf\n",
    "/home/paul/Zotero/storage/3IWJ9K27/Rabbani et al. - 2014 - The promise of whole-exome sequencing in medical g.pdf\n",
    "/home/paul/Zotero/storage/3JNN5N2W/Pengelly et al. - 2013 - A SNP profiling panel for sample tracking in whole.pdf\"\"\".splitlines()\n",
    "\n",
    "n_procs = min(n_procs, len(filelist))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "87c2ec6a-f68d-49f5-8e53-e1fa1bb140de",
   "metadata": {},
   "outputs": [],
   "source": [
    "# settings\n",
    "base_url = 'https://filesender.aarnet.edu.au/rest.php'\n",
    "default_transfer_days_valid = 21\n",
    "username = None\n",
    "apikey = None\n",
    "homepath = expanduser(\"~\")\n",
    "\n",
    "config = configparser.ConfigParser()\n",
    "config.read(homepath + '/.filesender/filesender.py.ini')\n",
    "if 'system' in config:\n",
    "    base_url = config['system'].get(\n",
    "        'base_url', 'https://filesender.aarnet.edu.au/rest.php')\n",
    "    default_transfer_days_valid = int(\n",
    "        config['system'].get('default_transfer_days_valid', 10))\n",
    "if 'user' in config:\n",
    "    username = config['user'].get('username')\n",
    "    apikey = config['user'].get('apikey')\n",
    "\n",
    "# if we have found these in the config file they become optional arguments\n",
    "if username is None:\n",
    "    requiredNamed.add_argument(\"-u\", \"--username\", required=True)\n",
    "else:\n",
    "    parser.add_argument(\"-u\", \"--username\")\n",
    "\n",
    "if apikey is None:\n",
    "    requiredNamed.add_argument(\"-a\", \"--apikey\", required=True)\n",
    "else:\n",
    "    parser.add_argument(\"-a\", \"--apikey\")\n",
    "\n",
    "requiredNamed.add_argument(\"-r\", \"--recipients\", required=True)\n",
    "# args = parser.parse_args(\"-p -r paul.wang@sahmri.com,paul.wang@sa-genomics.com.au -v -s subject -m messagebody\".split(\" \") + filelist)\n",
    "args = parser.parse_args(\"-p -r paul.wang@sahmri.com -v -s subject -m messagebody\".split(\" \") + filelist)\n",
    "debug = args.verbose\n",
    "progress = args.progress\n",
    "insecure = args.insecure\n",
    "skip_email = args.skip_email\n",
    "\n",
    "\n",
    "if args.username is not None:\n",
    "    username = args.username\n",
    "\n",
    "if args.apikey is not None:\n",
    "    apikey = args.apikey\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "b1223d36-74c3-467d-90da-ad65c443f244",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'url': 'https://filesender.aarnet.edu.au/',\n",
       " 'name': 'AARNet FileSender',\n",
       " 'logon_url': 'https://cloudstor.aarnet.edu.au/simplesaml/module.php/core/as_login.php?AuthId=default-sp&ReturnTo=__target__',\n",
       " 'upload_chunk_size': 12582912}"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "base_url = 'https://filesender.aarnet.edu.au/rest.php'\n",
    "response = requests.get(base_url+'/info', verify=True)\n",
    "response.json()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "88e83862-1da4-45c6-b96e-a92bd636f146",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "base_url          : https://filesender.aarnet.edu.au/rest.php\n",
      "username          : paul.wang@adelaide.edu.au\n",
      "apikey            : d1f80e881bb96605a1e903b6647330e7722b43bdd5385a09e3eb97ce9bd4ea19\n",
      "upload_chunk_size : 12582912 bytes\n",
      "recipients        : paul.wang@sahmri.com\n",
      "files             : /home/paul/Zotero/storage/28F8IG85/Lokody - 2013 - Cancer genomics Signature analysis suggests cance.pdf,/home/paul/Zotero/storage/29G6QV82/Usuyama et al. - 2014 - HapMuC somatic mutation calling using heterozygou.pdf,/home/paul/Zotero/storage/2BBJ65XW/Pleasance et al. - 2010 - A small cell lung cancer genome reports complex to.pdf,/home/paul/Zotero/storage/2CR2WGUC/Taub et al. - 2010 - Overcoming bias and systematic errors in next gene.pdf,/home/paul/Zotero/storage/2DRPFWV8/Kim et al. - 2016 - Somatic ERCC2 mutations are associated with a dist.pdf,/home/paul/Zotero/storage/2FXQNZ3S/Alter et al. - 2000 - Singular value decomposition for genome-wide expre.pdf,/home/paul/Zotero/storage/2GTA3FY6/Garcia-Prieto et al. - 2020 - The consequences of variant calling decisions in s.pdf,/home/paul/Zotero/storage/2JRP4NTD/Kim et al. - 2018 - Strelka2 fast and accurate calling of germline an.pdf,/home/paul/Zotero/storage/2P6RPSFQ/Pellegrino et al. - 2018 - High-throughput single-cell DNA sequencing of acut.pdf,/home/paul/Zotero/storage/2QNXMWPI/Kandoth et al. - 2013 - Mutational landscape and significance across 12 ma.pdf,/home/paul/Zotero/storage/2S6DPCZ7/Liu et al. - 2016 - Joint detection of copy number variations in paren.pdf,/home/paul/Zotero/storage/2UA83RGN/Consortium - 2012 - An integrated map of genetic variation from 1,092 .pdf,/home/paul/Zotero/storage/333VUI79/Zhu-SII09.pdf,/home/paul/Zotero/storage/34C3ERMA/Bersanelli et al. - 2016 - Methods for the integration of multi-omics data m.pdf,/home/paul/Zotero/storage/3BGPU5KB/Hansen et al. - 2012 - BSmooth from whole genome bisulfite sequencing re.pdf,/home/paul/Zotero/storage/3CFDCRFR/O'Brien et al. - 2010 - Gremlin An Interactive Visualization Model for An.pdf,/home/paul/Zotero/storage/3HBZH2XM/gim2014178a.pdf,/home/paul/Zotero/storage/3HMKCBF3/Parisi et al. - 2014 - Ranking and combining multiple predictors without .pdf,/home/paul/Zotero/storage/3IWJ9K27/Rabbani et al. - 2014 - The promise of whole-exome sequencing in medical g.pdf,/home/paul/Zotero/storage/3JNN5N2W/Pengelly et al. - 2013 - A SNP profiling panel for sample tracking in whole.pdf\n",
      "insecure          : False\n"
     ]
    }
   ],
   "source": [
    "\n",
    "# configs\n",
    "try:\n",
    "    response = requests.get(base_url+'/info', verify=True)\n",
    "except requests.exceptions.SSLError as exc:\n",
    "    if not insecure:\n",
    "        print('Error: the SSL certificate of the server you are connecting to cannot be verified:')\n",
    "        print(exc)\n",
    "        print('For more information, please refer to https://www.digicert.com/ssl/. If you are absolutely certain of the identity of the server you are connecting to, you can use the --insecure flag to bypass this warning. Exiting...')\n",
    "        sys.exit(1)\n",
    "    elif insecure:\n",
    "        print('Warning: Error: the SSL certificate of the server you are connecting to cannot be verified:')\n",
    "        print(exc)\n",
    "        print('Running with --insecure flag, ignoring warning...')\n",
    "        response = requests.get(base_url+'/info', verify=False)\n",
    "upload_chunk_size = response.json()['upload_chunk_size']\n",
    "\n",
    "if debug:\n",
    "    print('base_url          : '+base_url)\n",
    "    print('username          : '+username)\n",
    "    print('apikey            : '+apikey)\n",
    "    print('upload_chunk_size : '+str(upload_chunk_size)+' bytes')\n",
    "    print('recipients        : '+args.recipients)\n",
    "    print('files             : '+','.join(args.files))\n",
    "    print('insecure          : '+str(insecure))\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dad06f94-4485-47f9-a7ad-56a442dd0a4d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "aa0fe0d8-c223-4be9-a427-b548080b8565",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "postTransfer\n",
      "['paul.wang@sahmri.com']\n"
     ]
    }
   ],
   "source": [
    "\n",
    "# postTransfer\n",
    "if debug:\n",
    "    print('postTransfer')\n",
    "\n",
    "files = {}\n",
    "filesTransfer = []\n",
    "for f in args.files:\n",
    "    fn_abs = os.path.abspath(f)\n",
    "    fn = os.path.basename(fn_abs)\n",
    "    size = os.path.getsize(fn_abs)\n",
    "\n",
    "    files[fn+':'+str(size)] = {\n",
    "        'name': fn,\n",
    "        'size': size,\n",
    "        'path': fn_abs\n",
    "    }\n",
    "    filesTransfer.append({'name': fn, 'size': size})\n",
    "\n",
    "troptions = {'get_a_link': skip_email}\n",
    "\n",
    "# sort by decreasing file size\n",
    "filesTransfer = sorted(filesTransfer, key=lambda x: x[\"size\"], reverse=True)\n",
    "\n",
    "transfer = postTransfer(username,\n",
    "                        filesTransfer,\n",
    "                        args.recipients,\n",
    "                        subject=args.subject,\n",
    "                        message=args.message,\n",
    "                        expires=None,\n",
    "                        options=troptions)['created']\n",
    "# print(transfer)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d4ae0936-fa1c-42d7-9e2f-5bfe881c617e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "b43af6e1-1651-4abc-a297-280c1998a4a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# transfer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "990c57b8-eba2-4355-9153-e0a622319249",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "b1018a71-9a61-46b8-b9cf-e3f60304b1b4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "putChunks: /home/paul/Zotero/storage/2QNXMWPI/Kandoth et al. - 2013 - Mutational landscape and significance across 12 ma.pdf\n",
      "Uploading: /home/paul/Zotero/storage/2QNXMWPI/Kandoth et al. - 2013 - Mutational landscape and significance across 12 ma.pdf 0-8826624 0%\n",
      "uploaded 1 chunks\n",
      "fileComplete: /home/paul/Zotero/storage/2QNXMWPI/Kandoth et al. - 2013 - Mutational landscape and significance across 12 ma.pdf\n",
      "Uploading: /home/paul/Zotero/storage/2QNXMWPI/Kandoth et al. - 2013 - Mutational landscape and significance across 12 ma.pdf 627990 100%\n",
      "putChunks: /home/paul/Zotero/storage/3HBZH2XM/gim2014178a.pdf\n",
      "Uploading: /home/paul/Zotero/storage/3HBZH2XM/gim2014178a.pdf 0-5283692 0%\n",
      "uploaded 1 chunks\n",
      "fileComplete: /home/paul/Zotero/storage/3HBZH2XM/gim2014178a.pdf\n",
      "Uploading: /home/paul/Zotero/storage/3HBZH2XM/gim2014178a.pdf 627990 100%\n",
      "putChunks: /home/paul/Zotero/storage/2P6RPSFQ/Pellegrino et al. - 2018 - High-throughput single-cell DNA sequencing of acut.pdf\n",
      "Uploading: /home/paul/Zotero/storage/2P6RPSFQ/Pellegrino et al. - 2018 - High-throughput single-cell DNA sequencing of acut.pdf 0-3988467 0%\n",
      "uploaded 1 chunks\n",
      "fileComplete: /home/paul/Zotero/storage/2P6RPSFQ/Pellegrino et al. - 2018 - High-throughput single-cell DNA sequencing of acut.pdf\n",
      "Uploading: /home/paul/Zotero/storage/2P6RPSFQ/Pellegrino et al. - 2018 - High-throughput single-cell DNA sequencing of acut.pdf 627990 100%\n",
      "putChunks: /home/paul/Zotero/storage/2GTA3FY6/Garcia-Prieto et al. - 2020 - The consequences of variant calling decisions in s.pdf\n",
      "Uploading: /home/paul/Zotero/storage/2GTA3FY6/Garcia-Prieto et al. - 2020 - The consequences of variant calling decisions in s.pdf 0-3702595 0%\n",
      "uploaded 1 chunks\n",
      "fileComplete: /home/paul/Zotero/storage/2GTA3FY6/Garcia-Prieto et al. - 2020 - The consequences of variant calling decisions in s.pdf\n",
      "Uploading: /home/paul/Zotero/storage/2GTA3FY6/Garcia-Prieto et al. - 2020 - The consequences of variant calling decisions in s.pdf 627990 100%\n",
      "putChunks: /home/paul/Zotero/storage/2JRP4NTD/Kim et al. - 2018 - Strelka2 fast and accurate calling of germline an.pdf\n",
      "Uploading: /home/paul/Zotero/storage/2JRP4NTD/Kim et al. - 2018 - Strelka2 fast and accurate calling of germline an.pdf 0-2003929 0%\n",
      "uploaded 1 chunks\n",
      "fileComplete: /home/paul/Zotero/storage/2JRP4NTD/Kim et al. - 2018 - Strelka2 fast and accurate calling of germline an.pdf\n",
      "Uploading: /home/paul/Zotero/storage/2JRP4NTD/Kim et al. - 2018 - Strelka2 fast and accurate calling of germline an.pdf 627990 100%\n",
      "putChunks: /home/paul/Zotero/storage/2DRPFWV8/Kim et al. - 2016 - Somatic ERCC2 mutations are associated with a dist.pdf\n",
      "Uploading: /home/paul/Zotero/storage/2DRPFWV8/Kim et al. - 2016 - Somatic ERCC2 mutations are associated with a dist.pdf 0-1484048 0%\n",
      "uploaded 1 chunks\n",
      "fileComplete: /home/paul/Zotero/storage/2DRPFWV8/Kim et al. - 2016 - Somatic ERCC2 mutations are associated with a dist.pdf\n",
      "Uploading: /home/paul/Zotero/storage/2DRPFWV8/Kim et al. - 2016 - Somatic ERCC2 mutations are associated with a dist.pdf 627990 100%\n",
      "putChunks: /home/paul/Zotero/storage/2BBJ65XW/Pleasance et al. - 2010 - A small cell lung cancer genome reports complex to.pdf\n",
      "Uploading: /home/paul/Zotero/storage/2BBJ65XW/Pleasance et al. - 2010 - A small cell lung cancer genome reports complex to.pdf 0-1425635 0%\n",
      "uploaded 1 chunks\n",
      "fileComplete: /home/paul/Zotero/storage/2BBJ65XW/Pleasance et al. - 2010 - A small cell lung cancer genome reports complex to.pdf\n",
      "Uploading: /home/paul/Zotero/storage/2BBJ65XW/Pleasance et al. - 2010 - A small cell lung cancer genome reports complex to.pdf 627990 100%\n",
      "putChunks: /home/paul/Zotero/storage/2FXQNZ3S/Alter et al. - 2000 - Singular value decomposition for genome-wide expre.pdf\n",
      "Uploading: /home/paul/Zotero/storage/2FXQNZ3S/Alter et al. - 2000 - Singular value decomposition for genome-wide expre.pdf 0-1337907 0%\n",
      "uploaded 1 chunks\n",
      "fileComplete: /home/paul/Zotero/storage/2FXQNZ3S/Alter et al. - 2000 - Singular value decomposition for genome-wide expre.pdf\n",
      "Uploading: /home/paul/Zotero/storage/2FXQNZ3S/Alter et al. - 2000 - Singular value decomposition for genome-wide expre.pdf 627990 100%\n",
      "putChunks: /home/paul/Zotero/storage/3CFDCRFR/O'Brien et al. - 2010 - Gremlin An Interactive Visualization Model for An.pdf\n",
      "Uploading: /home/paul/Zotero/storage/3CFDCRFR/O'Brien et al. - 2010 - Gremlin An Interactive Visualization Model for An.pdf 0-1125508 0%\n",
      "uploaded 1 chunks\n",
      "fileComplete: /home/paul/Zotero/storage/3CFDCRFR/O'Brien et al. - 2010 - Gremlin An Interactive Visualization Model for An.pdf\n",
      "Uploading: /home/paul/Zotero/storage/3CFDCRFR/O'Brien et al. - 2010 - Gremlin An Interactive Visualization Model for An.pdf 627990 100%\n",
      "putChunks: /home/paul/Zotero/storage/2UA83RGN/Consortium - 2012 - An integrated map of genetic variation from 1,092 .pdf\n",
      "Uploading: /home/paul/Zotero/storage/2UA83RGN/Consortium - 2012 - An integrated map of genetic variation from 1,092 .pdf 0-1071963 0%\n",
      "uploaded 1 chunks\n",
      "fileComplete: /home/paul/Zotero/storage/2UA83RGN/Consortium - 2012 - An integrated map of genetic variation from 1,092 .pdf\n",
      "Uploading: /home/paul/Zotero/storage/2UA83RGN/Consortium - 2012 - An integrated map of genetic variation from 1,092 .pdf 627990 100%\n",
      "putChunks: /home/paul/Zotero/storage/3BGPU5KB/Hansen et al. - 2012 - BSmooth from whole genome bisulfite sequencing re.pdf\n",
      "Uploading: /home/paul/Zotero/storage/3BGPU5KB/Hansen et al. - 2012 - BSmooth from whole genome bisulfite sequencing re.pdf 0-970830 0%\n",
      "uploaded 1 chunks\n",
      "fileComplete: /home/paul/Zotero/storage/3BGPU5KB/Hansen et al. - 2012 - BSmooth from whole genome bisulfite sequencing re.pdf\n",
      "Uploading: /home/paul/Zotero/storage/3BGPU5KB/Hansen et al. - 2012 - BSmooth from whole genome bisulfite sequencing re.pdf 627990 100%\n",
      "putChunks: /home/paul/Zotero/storage/3IWJ9K27/Rabbani et al. - 2014 - The promise of whole-exome sequencing in medical g.pdf\n",
      "Uploading: /home/paul/Zotero/storage/3IWJ9K27/Rabbani et al. - 2014 - The promise of whole-exome sequencing in medical g.pdf 0-836844 0%\n",
      "uploaded 1 chunks\n",
      "fileComplete: /home/paul/Zotero/storage/3IWJ9K27/Rabbani et al. - 2014 - The promise of whole-exome sequencing in medical g.pdf\n",
      "Uploading: /home/paul/Zotero/storage/3IWJ9K27/Rabbani et al. - 2014 - The promise of whole-exome sequencing in medical g.pdf 627990 100%\n",
      "putChunks: /home/paul/Zotero/storage/3HMKCBF3/Parisi et al. - 2014 - Ranking and combining multiple predictors without .pdf\n",
      "Uploading: /home/paul/Zotero/storage/3HMKCBF3/Parisi et al. - 2014 - Ranking and combining multiple predictors without .pdf 0-699461 0%\n",
      "uploaded 1 chunks\n",
      "fileComplete: /home/paul/Zotero/storage/3HMKCBF3/Parisi et al. - 2014 - Ranking and combining multiple predictors without .pdf\n",
      "Uploading: /home/paul/Zotero/storage/3HMKCBF3/Parisi et al. - 2014 - Ranking and combining multiple predictors without .pdf 627990 100%\n",
      "putChunks: /home/paul/Zotero/storage/2CR2WGUC/Taub et al. - 2010 - Overcoming bias and systematic errors in next gene.pdf\n",
      "Uploading: /home/paul/Zotero/storage/2CR2WGUC/Taub et al. - 2010 - Overcoming bias and systematic errors in next gene.pdf 0-698740 0%\n",
      "uploaded 1 chunks\n",
      "fileComplete: /home/paul/Zotero/storage/2CR2WGUC/Taub et al. - 2010 - Overcoming bias and systematic errors in next gene.pdf\n",
      "Uploading: /home/paul/Zotero/storage/2CR2WGUC/Taub et al. - 2010 - Overcoming bias and systematic errors in next gene.pdf 627990 100%\n",
      "putChunks: /home/paul/Zotero/storage/34C3ERMA/Bersanelli et al. - 2016 - Methods for the integration of multi-omics data m.pdf\n",
      "Uploading: /home/paul/Zotero/storage/34C3ERMA/Bersanelli et al. - 2016 - Methods for the integration of multi-omics data m.pdf 0-654800 0%\n",
      "uploaded 1 chunks\n",
      "fileComplete: /home/paul/Zotero/storage/34C3ERMA/Bersanelli et al. - 2016 - Methods for the integration of multi-omics data m.pdf\n",
      "Uploading: /home/paul/Zotero/storage/34C3ERMA/Bersanelli et al. - 2016 - Methods for the integration of multi-omics data m.pdf 627990 100%\n",
      "putChunks: /home/paul/Zotero/storage/3JNN5N2W/Pengelly et al. - 2013 - A SNP profiling panel for sample tracking in whole.pdf\n",
      "Uploading: /home/paul/Zotero/storage/3JNN5N2W/Pengelly et al. - 2013 - A SNP profiling panel for sample tracking in whole.pdf 0-627990 0%\n",
      "uploaded 1 chunks\n",
      "fileComplete: /home/paul/Zotero/storage/3JNN5N2W/Pengelly et al. - 2013 - A SNP profiling panel for sample tracking in whole.pdf\n",
      "Uploading: /home/paul/Zotero/storage/3JNN5N2W/Pengelly et al. - 2013 - A SNP profiling panel for sample tracking in whole.pdf 627990 100%\n",
      "putChunks: /home/paul/Zotero/storage/29G6QV82/Usuyama et al. - 2014 - HapMuC somatic mutation calling using heterozygou.pdf\n",
      "Uploading: /home/paul/Zotero/storage/29G6QV82/Usuyama et al. - 2014 - HapMuC somatic mutation calling using heterozygou.pdf 0-543903 0%\n",
      "uploaded 1 chunks\n",
      "fileComplete: /home/paul/Zotero/storage/29G6QV82/Usuyama et al. - 2014 - HapMuC somatic mutation calling using heterozygou.pdf\n",
      "Uploading: /home/paul/Zotero/storage/29G6QV82/Usuyama et al. - 2014 - HapMuC somatic mutation calling using heterozygou.pdf 627990 100%\n",
      "putChunks: /home/paul/Zotero/storage/2S6DPCZ7/Liu et al. - 2016 - Joint detection of copy number variations in paren.pdf\n",
      "Uploading: /home/paul/Zotero/storage/2S6DPCZ7/Liu et al. - 2016 - Joint detection of copy number variations in paren.pdf 0-539632 0%\n",
      "uploaded 1 chunks\n",
      "fileComplete: /home/paul/Zotero/storage/2S6DPCZ7/Liu et al. - 2016 - Joint detection of copy number variations in paren.pdf\n",
      "Uploading: /home/paul/Zotero/storage/2S6DPCZ7/Liu et al. - 2016 - Joint detection of copy number variations in paren.pdf 627990 100%\n",
      "putChunks: /home/paul/Zotero/storage/333VUI79/Zhu-SII09.pdf\n",
      "Uploading: /home/paul/Zotero/storage/333VUI79/Zhu-SII09.pdf 0-444617 0%\n",
      "uploaded 1 chunks\n",
      "fileComplete: /home/paul/Zotero/storage/333VUI79/Zhu-SII09.pdf\n",
      "Uploading: /home/paul/Zotero/storage/333VUI79/Zhu-SII09.pdf 627990 100%\n",
      "putChunks: /home/paul/Zotero/storage/28F8IG85/Lokody - 2013 - Cancer genomics Signature analysis suggests cance.pdf\n",
      "Uploading: /home/paul/Zotero/storage/28F8IG85/Lokody - 2013 - Cancer genomics Signature analysis suggests cance.pdf 0-360956 0%\n",
      "uploaded 1 chunks\n",
      "fileComplete: /home/paul/Zotero/storage/28F8IG85/Lokody - 2013 - Cancer genomics Signature analysis suggests cance.pdf\n",
      "Uploading: /home/paul/Zotero/storage/28F8IG85/Lokody - 2013 - Cancer genomics Signature analysis suggests cance.pdf 627990 100%\n",
      "11.58905577659607\n"
     ]
    }
   ],
   "source": [
    "t0 = time.time()\n",
    "\n",
    "n_procs = 1\n",
    "task = partial(upload_file, \n",
    "               transferData=transfer, \n",
    "               filesData=files,\n",
    "               upload_chunk_size=upload_chunk_size,\n",
    "               debug=debug)\n",
    "\n",
    "pool = Pool(n_procs)\n",
    "results = pool.map(task, transfer['files'])\n",
    "pool.close()\n",
    "\n",
    "t1 = time.time()\n",
    "print(t1-t0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "1cc3d723-f691-40d2-b366-9405fc0eddd1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "transferComplete\n",
      "Upload Complete\n"
     ]
    }
   ],
   "source": [
    "# transferComplete\n",
    "if debug:\n",
    "    print('transferComplete')\n",
    "res = transferComplete(transfer)\n",
    "if progress:\n",
    "    print('Upload Complete')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c1879a55-b673-48b4-be9b-408c17cebe73",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "87903fbc-2b26-4a3c-8844-d6b97f0faa5a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "id": "6a251a76-019a-4deb-a403-b1c45fd20164",
   "metadata": {},
   "outputs": [],
   "source": [
    "fout = open(\"./test/final_dump.json\", \"w\")\n",
    "fout.write(json.dumps(res))\n",
    "fout.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b1819e6e-7d52-4cb6-ae39-4bed37a52ee0",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "e965a93e-be42-44b3-9f2e-706417759885",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "FileSender upload\n",
      "transfer ID: 3137421\n",
      "uploaded by: paul.wang@adelaide.edu.au\n",
      "total size:  34.93 MB\n",
      "\n",
      "recipient:   paul.wang@sahmri.com\n",
      "D/L link:    https://filesender.aarnet.edu.au/?s=download&token=2c3d7a65-6408-4ad0-827f-cdd096d8cee5\n",
      "\n",
      "Files uploaded:\n",
      "\n",
      " 0: \"Kandoth et al. - 2013 - Mutational landscape and significance across 12 ma.pdf\"\n",
      "    8,826,624 bytes \n",
      "    https://filesender.aarnet.edu.au/download.php?token=2c3d7a65-6408-4ad0-827f-cdd096d8cee5&files_ids=21052951&files_ids=21052951\n",
      "\n",
      " 1: \"gim2014178a.pdf\"\n",
      "    5,283,692 bytes \n",
      "    https://filesender.aarnet.edu.au/download.php?token=2c3d7a65-6408-4ad0-827f-cdd096d8cee5&files_ids=21052956&files_ids=21052956\n",
      "\n",
      " 2: \"Pellegrino et al. - 2018 - High-throughput single-cell DNA sequencing of acut.pdf\"\n",
      "    3,988,467 bytes \n",
      "    https://filesender.aarnet.edu.au/download.php?token=2c3d7a65-6408-4ad0-827f-cdd096d8cee5&files_ids=21052961&files_ids=21052961\n",
      "\n",
      " 3: \"Garcia-Prieto et al. - 2020 - The consequences of variant calling decisions in s.pdf\"\n",
      "    3,702,595 bytes \n",
      "    https://filesender.aarnet.edu.au/download.php?token=2c3d7a65-6408-4ad0-827f-cdd096d8cee5&files_ids=21052966&files_ids=21052966\n",
      "\n",
      " 4: \"Kim et al. - 2018 - Strelka2 fast and accurate calling of germline an.pdf\"\n",
      "    2,003,929 bytes \n",
      "    https://filesender.aarnet.edu.au/download.php?token=2c3d7a65-6408-4ad0-827f-cdd096d8cee5&files_ids=21052971&files_ids=21052971\n",
      "\n",
      " 5: \"Kim et al. - 2016 - Somatic ERCC2 mutations are associated with a dist.pdf\"\n",
      "    1,484,048 bytes \n",
      "    https://filesender.aarnet.edu.au/download.php?token=2c3d7a65-6408-4ad0-827f-cdd096d8cee5&files_ids=21052976&files_ids=21052976\n",
      "\n",
      " 6: \"Pleasance et al. - 2010 - A small cell lung cancer genome reports complex to.pdf\"\n",
      "    1,425,635 bytes \n",
      "    https://filesender.aarnet.edu.au/download.php?token=2c3d7a65-6408-4ad0-827f-cdd096d8cee5&files_ids=21052981&files_ids=21052981\n",
      "\n",
      " 7: \"Alter et al. - 2000 - Singular value decomposition for genome-wide expre.pdf\"\n",
      "    1,337,907 bytes \n",
      "    https://filesender.aarnet.edu.au/download.php?token=2c3d7a65-6408-4ad0-827f-cdd096d8cee5&files_ids=21052986&files_ids=21052986\n",
      "\n",
      " 8: \"O'Brien et al. - 2010 - Gremlin An Interactive Visualization Model for An.pdf\"\n",
      "    1,125,508 bytes \n",
      "    https://filesender.aarnet.edu.au/download.php?token=2c3d7a65-6408-4ad0-827f-cdd096d8cee5&files_ids=21052991&files_ids=21052991\n",
      "\n",
      " 9: \"Consortium - 2012 - An integrated map of genetic variation from 1,092 .pdf\"\n",
      "    1,071,963 bytes \n",
      "    https://filesender.aarnet.edu.au/download.php?token=2c3d7a65-6408-4ad0-827f-cdd096d8cee5&files_ids=21052996&files_ids=21052996\n",
      "\n",
      "10: \"Hansen et al. - 2012 - BSmooth from whole genome bisulfite sequencing re.pdf\"\n",
      "    970,830 bytes \n",
      "    https://filesender.aarnet.edu.au/download.php?token=2c3d7a65-6408-4ad0-827f-cdd096d8cee5&files_ids=21053001&files_ids=21053001\n",
      "\n",
      "11: \"Rabbani et al. - 2014 - The promise of whole-exome sequencing in medical g.pdf\"\n",
      "    836,844 bytes \n",
      "    https://filesender.aarnet.edu.au/download.php?token=2c3d7a65-6408-4ad0-827f-cdd096d8cee5&files_ids=21053006&files_ids=21053006\n",
      "\n",
      "12: \"Parisi et al. - 2014 - Ranking and combining multiple predictors without .pdf\"\n",
      "    699,461 bytes \n",
      "    https://filesender.aarnet.edu.au/download.php?token=2c3d7a65-6408-4ad0-827f-cdd096d8cee5&files_ids=21053011&files_ids=21053011\n",
      "\n",
      "13: \"Taub et al. - 2010 - Overcoming bias and systematic errors in next gene.pdf\"\n",
      "    698,740 bytes \n",
      "    https://filesender.aarnet.edu.au/download.php?token=2c3d7a65-6408-4ad0-827f-cdd096d8cee5&files_ids=21053016&files_ids=21053016\n",
      "\n",
      "14: \"Bersanelli et al. - 2016 - Methods for the integration of multi-omics data m.pdf\"\n",
      "    654,800 bytes \n",
      "    https://filesender.aarnet.edu.au/download.php?token=2c3d7a65-6408-4ad0-827f-cdd096d8cee5&files_ids=21053021&files_ids=21053021\n",
      "\n",
      "15: \"Pengelly et al. - 2013 - A SNP profiling panel for sample tracking in whole.pdf\"\n",
      "    627,990 bytes \n",
      "    https://filesender.aarnet.edu.au/download.php?token=2c3d7a65-6408-4ad0-827f-cdd096d8cee5&files_ids=21053026&files_ids=21053026\n",
      "\n",
      "16: \"Usuyama et al. - 2014 - HapMuC somatic mutation calling using heterozygou.pdf\"\n",
      "    543,903 bytes \n",
      "    https://filesender.aarnet.edu.au/download.php?token=2c3d7a65-6408-4ad0-827f-cdd096d8cee5&files_ids=21053031&files_ids=21053031\n",
      "\n",
      "17: \"Liu et al. - 2016 - Joint detection of copy number variations in paren.pdf\"\n",
      "    539,632 bytes \n",
      "    https://filesender.aarnet.edu.au/download.php?token=2c3d7a65-6408-4ad0-827f-cdd096d8cee5&files_ids=21053036&files_ids=21053036\n",
      "\n",
      "18: \"Zhu-SII09.pdf\"\n",
      "    444,617 bytes \n",
      "    https://filesender.aarnet.edu.au/download.php?token=2c3d7a65-6408-4ad0-827f-cdd096d8cee5&files_ids=21053041&files_ids=21053041\n",
      "\n",
      "19: \"Lokody - 2013 - Cancer genomics Signature analysis suggests cance.pdf\"\n",
      "    360,956 bytes \n",
      "    https://filesender.aarnet.edu.au/download.php?token=2c3d7a65-6408-4ad0-827f-cdd096d8cee5&files_ids=21053046&files_ids=21053046\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(transfer_data_to_text(res))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4ecc9c01-2ec9-4bf3-9bfe-8abe97ba6193",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "251eb072-a00f-4daa-b29c-7c851efc64e1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'id': 3137421,\n",
       " 'userid': 76641,\n",
       " 'user_email': 'paul.wang@adelaide.edu.au',\n",
       " 'subject': 'subject',\n",
       " 'message': 'messagebody',\n",
       " 'created': {'raw': 1699503146, 'formatted': '09/11/2023'},\n",
       " 'expires': {'raw': 1701317535, 'formatted': '30/11/2023'},\n",
       " 'expiry_date_extension': 0,\n",
       " 'options': {'get_a_link': False,\n",
       "  'add_me_to_recipients': False,\n",
       "  'email_recipient_when_transfer_expires': True,\n",
       "  'encryption': False},\n",
       " 'salt': 'EKuz3wknSXnUeswS0k53rga8SoYgWnxl',\n",
       " 'roundtriptoken': '',\n",
       " 'files': [{'id': 21052951,\n",
       "   'transfer_id': 3137421,\n",
       "   'uid': 'c09c2587-73f5-4c30-a0f0-16d3dce2b4d9',\n",
       "   'name': 'Kandoth et al. - 2013 - Mutational landscape and significance across 12 ma.pdf',\n",
       "   'size': 8826624,\n",
       "   'sha1': None},\n",
       "  {'id': 21052956,\n",
       "   'transfer_id': 3137421,\n",
       "   'uid': '11693a50-836c-40c3-a8ce-de4b9385dbce',\n",
       "   'name': 'gim2014178a.pdf',\n",
       "   'size': 5283692,\n",
       "   'sha1': None},\n",
       "  {'id': 21052961,\n",
       "   'transfer_id': 3137421,\n",
       "   'uid': '45bc05fd-87d3-4621-8b73-3102192fcbc1',\n",
       "   'name': 'Pellegrino et al. - 2018 - High-throughput single-cell DNA sequencing of acut.pdf',\n",
       "   'size': 3988467,\n",
       "   'sha1': None},\n",
       "  {'id': 21052966,\n",
       "   'transfer_id': 3137421,\n",
       "   'uid': '37a778e2-b787-42f7-8817-9ca0032bf9a8',\n",
       "   'name': 'Garcia-Prieto et al. - 2020 - The consequences of variant calling decisions in s.pdf',\n",
       "   'size': 3702595,\n",
       "   'sha1': None},\n",
       "  {'id': 21052971,\n",
       "   'transfer_id': 3137421,\n",
       "   'uid': 'acdd3e4d-40fb-4457-bcf1-774e64a6b251',\n",
       "   'name': 'Kim et al. - 2018 - Strelka2 fast and accurate calling of germline an.pdf',\n",
       "   'size': 2003929,\n",
       "   'sha1': None},\n",
       "  {'id': 21052976,\n",
       "   'transfer_id': 3137421,\n",
       "   'uid': '239c7b95-4bf7-4398-9ef7-4dab7a8c1bca',\n",
       "   'name': 'Kim et al. - 2016 - Somatic ERCC2 mutations are associated with a dist.pdf',\n",
       "   'size': 1484048,\n",
       "   'sha1': None},\n",
       "  {'id': 21052981,\n",
       "   'transfer_id': 3137421,\n",
       "   'uid': '05ec54f1-82f6-4585-a1bb-77422f06f540',\n",
       "   'name': 'Pleasance et al. - 2010 - A small cell lung cancer genome reports complex to.pdf',\n",
       "   'size': 1425635,\n",
       "   'sha1': None},\n",
       "  {'id': 21052986,\n",
       "   'transfer_id': 3137421,\n",
       "   'uid': 'abf76af7-2740-4981-a284-36f80e31e1fd',\n",
       "   'name': 'Alter et al. - 2000 - Singular value decomposition for genome-wide expre.pdf',\n",
       "   'size': 1337907,\n",
       "   'sha1': None},\n",
       "  {'id': 21052991,\n",
       "   'transfer_id': 3137421,\n",
       "   'uid': '11c716a8-a73f-4755-af09-e25b0c3ea064',\n",
       "   'name': \"O'Brien et al. - 2010 - Gremlin An Interactive Visualization Model for An.pdf\",\n",
       "   'size': 1125508,\n",
       "   'sha1': None},\n",
       "  {'id': 21052996,\n",
       "   'transfer_id': 3137421,\n",
       "   'uid': 'e135350e-e9cf-4d59-845b-db4a11129826',\n",
       "   'name': 'Consortium - 2012 - An integrated map of genetic variation from 1,092 .pdf',\n",
       "   'size': 1071963,\n",
       "   'sha1': None},\n",
       "  {'id': 21053001,\n",
       "   'transfer_id': 3137421,\n",
       "   'uid': '38938666-5ee7-41c2-b70b-ffc0b6b03f47',\n",
       "   'name': 'Hansen et al. - 2012 - BSmooth from whole genome bisulfite sequencing re.pdf',\n",
       "   'size': 970830,\n",
       "   'sha1': None},\n",
       "  {'id': 21053006,\n",
       "   'transfer_id': 3137421,\n",
       "   'uid': '8c70fd36-3c9d-49a8-b214-f18076dac7a6',\n",
       "   'name': 'Rabbani et al. - 2014 - The promise of whole-exome sequencing in medical g.pdf',\n",
       "   'size': 836844,\n",
       "   'sha1': None},\n",
       "  {'id': 21053011,\n",
       "   'transfer_id': 3137421,\n",
       "   'uid': '7e6098a8-00db-408e-9ae0-a2f9206d3734',\n",
       "   'name': 'Parisi et al. - 2014 - Ranking and combining multiple predictors without .pdf',\n",
       "   'size': 699461,\n",
       "   'sha1': None},\n",
       "  {'id': 21053016,\n",
       "   'transfer_id': 3137421,\n",
       "   'uid': 'ea96fb24-bc7d-47ff-88be-6710b8ee5844',\n",
       "   'name': 'Taub et al. - 2010 - Overcoming bias and systematic errors in next gene.pdf',\n",
       "   'size': 698740,\n",
       "   'sha1': None},\n",
       "  {'id': 21053021,\n",
       "   'transfer_id': 3137421,\n",
       "   'uid': 'd8a29d4d-0d0b-458f-b295-dcac0dd7cb48',\n",
       "   'name': 'Bersanelli et al. - 2016 - Methods for the integration of multi-omics data m.pdf',\n",
       "   'size': 654800,\n",
       "   'sha1': None},\n",
       "  {'id': 21053026,\n",
       "   'transfer_id': 3137421,\n",
       "   'uid': 'ca9ba441-f0aa-4989-997b-0422f073d6b9',\n",
       "   'name': 'Pengelly et al. - 2013 - A SNP profiling panel for sample tracking in whole.pdf',\n",
       "   'size': 627990,\n",
       "   'sha1': None},\n",
       "  {'id': 21053031,\n",
       "   'transfer_id': 3137421,\n",
       "   'uid': '0d999915-4b89-44b5-be28-3814c00775de',\n",
       "   'name': 'Usuyama et al. - 2014 - HapMuC somatic mutation calling using heterozygou.pdf',\n",
       "   'size': 543903,\n",
       "   'sha1': None},\n",
       "  {'id': 21053036,\n",
       "   'transfer_id': 3137421,\n",
       "   'uid': '20477896-90f0-41df-8699-53557e6db3b1',\n",
       "   'name': 'Liu et al. - 2016 - Joint detection of copy number variations in paren.pdf',\n",
       "   'size': 539632,\n",
       "   'sha1': None},\n",
       "  {'id': 21053041,\n",
       "   'transfer_id': 3137421,\n",
       "   'uid': '8f74cf01-e138-47a2-b11e-ea484fd5289f',\n",
       "   'name': 'Zhu-SII09.pdf',\n",
       "   'size': 444617,\n",
       "   'sha1': None},\n",
       "  {'id': 21053046,\n",
       "   'transfer_id': 3137421,\n",
       "   'uid': '7b76a793-1c7e-495c-b0aa-d3de1525f800',\n",
       "   'name': 'Lokody - 2013 - Cancer genomics Signature analysis suggests cance.pdf',\n",
       "   'size': 360956,\n",
       "   'sha1': None}],\n",
       " 'recipients': [{'id': 5455206,\n",
       "   'transfer_id': 3137421,\n",
       "   'token': '2c3d7a65-6408-4ad0-827f-cdd096d8cee5',\n",
       "   'email': 'paul.wang@sahmri.com',\n",
       "   'created': {'raw': 1699503147, 'formatted': '09/11/2023'},\n",
       "   'last_activity': None,\n",
       "   'options': None,\n",
       "   'download_url': 'https://filesender.aarnet.edu.au/?s=download&token=2c3d7a65-6408-4ad0-827f-cdd096d8cee5',\n",
       "   'errors': []}]}"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d54165bc-ebd5-42ac-9942-161df38ecdca",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "550d3eed-1f48-44e2-a06d-ac6d6bc2f52a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b899b921-708e-4b4e-95d2-b40483e24134",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4a4e12f6-4479-4343-a7fe-7cf3c6fc0481",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cad2d6ee-563f-45c7-afe2-b0a8e2ee38f6",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "id": "d05200b0-8eb4-4a03-b8f9-7b9119e735d1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "FileSender upload\n",
      "transfer ID: 3137356\n",
      "uploaded by: paul.wang@adelaide.edu.au\n",
      "total size:  34.93 MB\n",
      "\n",
      "recipient:   paul.wang@sahmri.com\n",
      "D/L link:    https://filesender.aarnet.edu.au/?s=download&token=3aea5b1c-4ce6-4328-8d9a-1ccde0d2bf20\n",
      "\n",
      "Files uploaded:\n",
      "\n",
      " 0: \"Kandoth et al. - 2013 - Mutational landscape and significance across 12 ma.pdf\"\n",
      "    8,826,624 bytes \n",
      "    url: https://filesender.aarnet.edu.au/download.php?token=3aea5b1c-4ce6-4328-8d9a-1ccde0d2bf20&files_ids=21052436&files_ids=21052436\n",
      "\n",
      " 1: \"gim2014178a.pdf\"\n",
      "    5,283,692 bytes \n",
      "    url: https://filesender.aarnet.edu.au/download.php?token=3aea5b1c-4ce6-4328-8d9a-1ccde0d2bf20&files_ids=21052441&files_ids=21052441\n",
      "\n",
      " 2: \"Pellegrino et al. - 2018 - High-throughput single-cell DNA sequencing of acut.pdf\"\n",
      "    3,988,467 bytes \n",
      "    url: https://filesender.aarnet.edu.au/download.php?token=3aea5b1c-4ce6-4328-8d9a-1ccde0d2bf20&files_ids=21052446&files_ids=21052446\n",
      "\n",
      " 3: \"Garcia-Prieto et al. - 2020 - The consequences of variant calling decisions in s.pdf\"\n",
      "    3,702,595 bytes \n",
      "    url: https://filesender.aarnet.edu.au/download.php?token=3aea5b1c-4ce6-4328-8d9a-1ccde0d2bf20&files_ids=21052451&files_ids=21052451\n",
      "\n",
      " 4: \"Kim et al. - 2018 - Strelka2 fast and accurate calling of germline an.pdf\"\n",
      "    2,003,929 bytes \n",
      "    url: https://filesender.aarnet.edu.au/download.php?token=3aea5b1c-4ce6-4328-8d9a-1ccde0d2bf20&files_ids=21052456&files_ids=21052456\n",
      "\n",
      " 5: \"Kim et al. - 2016 - Somatic ERCC2 mutations are associated with a dist.pdf\"\n",
      "    1,484,048 bytes \n",
      "    url: https://filesender.aarnet.edu.au/download.php?token=3aea5b1c-4ce6-4328-8d9a-1ccde0d2bf20&files_ids=21052461&files_ids=21052461\n",
      "\n",
      " 6: \"Pleasance et al. - 2010 - A small cell lung cancer genome reports complex to.pdf\"\n",
      "    1,425,635 bytes \n",
      "    url: https://filesender.aarnet.edu.au/download.php?token=3aea5b1c-4ce6-4328-8d9a-1ccde0d2bf20&files_ids=21052466&files_ids=21052466\n",
      "\n",
      " 7: \"Alter et al. - 2000 - Singular value decomposition for genome-wide expre.pdf\"\n",
      "    1,337,907 bytes \n",
      "    url: https://filesender.aarnet.edu.au/download.php?token=3aea5b1c-4ce6-4328-8d9a-1ccde0d2bf20&files_ids=21052471&files_ids=21052471\n",
      "\n",
      " 8: \"O'Brien et al. - 2010 - Gremlin An Interactive Visualization Model for An.pdf\"\n",
      "    1,125,508 bytes \n",
      "    url: https://filesender.aarnet.edu.au/download.php?token=3aea5b1c-4ce6-4328-8d9a-1ccde0d2bf20&files_ids=21052476&files_ids=21052476\n",
      "\n",
      " 9: \"Consortium - 2012 - An integrated map of genetic variation from 1,092 .pdf\"\n",
      "    1,071,963 bytes \n",
      "    url: https://filesender.aarnet.edu.au/download.php?token=3aea5b1c-4ce6-4328-8d9a-1ccde0d2bf20&files_ids=21052481&files_ids=21052481\n",
      "\n",
      "10: \"Hansen et al. - 2012 - BSmooth from whole genome bisulfite sequencing re.pdf\"\n",
      "    970,830 bytes \n",
      "    url: https://filesender.aarnet.edu.au/download.php?token=3aea5b1c-4ce6-4328-8d9a-1ccde0d2bf20&files_ids=21052486&files_ids=21052486\n",
      "\n",
      "11: \"Rabbani et al. - 2014 - The promise of whole-exome sequencing in medical g.pdf\"\n",
      "    836,844 bytes \n",
      "    url: https://filesender.aarnet.edu.au/download.php?token=3aea5b1c-4ce6-4328-8d9a-1ccde0d2bf20&files_ids=21052491&files_ids=21052491\n",
      "\n",
      "12: \"Parisi et al. - 2014 - Ranking and combining multiple predictors without .pdf\"\n",
      "    699,461 bytes \n",
      "    url: https://filesender.aarnet.edu.au/download.php?token=3aea5b1c-4ce6-4328-8d9a-1ccde0d2bf20&files_ids=21052496&files_ids=21052496\n",
      "\n",
      "13: \"Taub et al. - 2010 - Overcoming bias and systematic errors in next gene.pdf\"\n",
      "    698,740 bytes \n",
      "    url: https://filesender.aarnet.edu.au/download.php?token=3aea5b1c-4ce6-4328-8d9a-1ccde0d2bf20&files_ids=21052501&files_ids=21052501\n",
      "\n",
      "14: \"Bersanelli et al. - 2016 - Methods for the integration of multi-omics data m.pdf\"\n",
      "    654,800 bytes \n",
      "    url: https://filesender.aarnet.edu.au/download.php?token=3aea5b1c-4ce6-4328-8d9a-1ccde0d2bf20&files_ids=21052506&files_ids=21052506\n",
      "\n",
      "15: \"Pengelly et al. - 2013 - A SNP profiling panel for sample tracking in whole.pdf\"\n",
      "    627,990 bytes \n",
      "    url: https://filesender.aarnet.edu.au/download.php?token=3aea5b1c-4ce6-4328-8d9a-1ccde0d2bf20&files_ids=21052511&files_ids=21052511\n",
      "\n",
      "16: \"Usuyama et al. - 2014 - HapMuC somatic mutation calling using heterozygou.pdf\"\n",
      "    543,903 bytes \n",
      "    url: https://filesender.aarnet.edu.au/download.php?token=3aea5b1c-4ce6-4328-8d9a-1ccde0d2bf20&files_ids=21052516&files_ids=21052516\n",
      "\n",
      "17: \"Liu et al. - 2016 - Joint detection of copy number variations in paren.pdf\"\n",
      "    539,632 bytes \n",
      "    url: https://filesender.aarnet.edu.au/download.php?token=3aea5b1c-4ce6-4328-8d9a-1ccde0d2bf20&files_ids=21052521&files_ids=21052521\n",
      "\n",
      "18: \"Zhu-SII09.pdf\"\n",
      "    444,617 bytes \n",
      "    url: https://filesender.aarnet.edu.au/download.php?token=3aea5b1c-4ce6-4328-8d9a-1ccde0d2bf20&files_ids=21052526&files_ids=21052526\n",
      "\n",
      "19: \"Lokody - 2013 - Cancer genomics Signature analysis suggests cance.pdf\"\n",
      "    360,956 bytes \n",
      "    url: https://filesender.aarnet.edu.au/download.php?token=3aea5b1c-4ce6-4328-8d9a-1ccde0d2bf20&files_ids=21052531&files_ids=21052531\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# text output\n",
    "\n",
    "tdata = res\n",
    "\n",
    "total_size = 0\n",
    "for f in tdata[\"files\"]:\n",
    "    total_size += f[\"size\"]\n",
    "size_unit = \"B\"\n",
    "if total_size > 1024**3:\n",
    "    size_unit = \"GB\"\n",
    "    total_size = total_size/(1024**3)\n",
    "elif total_size > 1024**2:\n",
    "    size_unit = \"MB\"\n",
    "    total_size = total_size/(1024**2)\n",
    "elif total_size > 1024:\n",
    "    size_unit = \"kB\"\n",
    "    total_size = total_size/(1024)\n",
    "\n",
    "size_str = f\"{total_size:.2f} {size_unit}\"\n",
    "\n",
    "\n",
    "url = res[\"recipients\"][0][\"download_url\"]\n",
    "\n",
    "report_txt = f\"\"\"FileSender upload\n",
    "transfer ID: {tdata[\"id\"]}\n",
    "uploaded by: {tdata[\"user_email\"]}\n",
    "total size:  {size_str}\n",
    "\n",
    "recipient:   {res[\"recipients\"][0][\"email\"]}\n",
    "D/L link:    {url}\n",
    "\n",
    "Files uploaded:\n",
    "\"\"\"\n",
    "\n",
    "for fidx, f in enumerate(res[\"files\"]):\n",
    "    url = f\"https://filesender.aarnet.edu.au/download.php?token={res['recipients'][0]['token']}&files_ids={f['id']}\"\n",
    "    ct_str = f\"{fidx}\".rjust( len(str(len(res[\"files\"]))) )\n",
    "    report_txt += f\"\"\"\n",
    "{ct_str}: \"{f[\"name\"]}\"\n",
    "{\" \"*len(ct_str)}  {f[\"size\"]:,} bytes \n",
    "{\" \"*len(ct_str)}  url: {url}&files_ids={f[\"id\"]}\n",
    "\"\"\"\n",
    "\n",
    "# print(report_txt)\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "id": "c9c18789-adff-4d91-a3b2-b7d1e715a48a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'id': 5455011,\n",
       "  'transfer_id': 3137241,\n",
       "  'token': 'd77029ce-42d3-4e6f-ab7a-8afd9edf2504',\n",
       "  'email': '',\n",
       "  'created': {'raw': 1699500252, 'formatted': '09/11/2023'},\n",
       "  'last_activity': None,\n",
       "  'options': None,\n",
       "  'download_url': 'https://filesender.aarnet.edu.au/?s=download&token=d77029ce-42d3-4e6f-ab7a-8afd9edf2504',\n",
       "  'errors': []}]"
      ]
     },
     "execution_count": 105,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res[\"recipients\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "84de8741-0b81-4559-aedb-29b108ef9b72",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4638be59-16a6-44d5-b3c4-fd470efc93b3",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ae04be7d-5438-485b-8e70-5a8c24c8a5f1",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e95d11f5-6f7c-4894-ae04-0e08f9d40618",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9efb9130-421e-4c52-80f7-0d25c8eb4ae7",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4c486636-d378-436f-ac05-a224b27feda3",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "befc81c9-d5b1-4194-9396-c74aaf3040dc",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "raw",
   "id": "e6641376-2357-42ee-bcdd-06c1b0b4756d",
   "metadata": {},
   "source": [
    "tid = 3136171\n",
    "\n",
    "call(\"get\", \n",
    "     f\"/transfer/{tid}\",\n",
    "     {}, None, None, {}) \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "034b19e8-3f54-42ba-a596-9b4444c0ba08",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2c681025-50bc-47d0-9e3e-33ea5620e7b5",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b5f60f2a-a203-4006-91a3-de462b87e228",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "96bb446b-1c61-4efe-b04b-4e6b73dee612",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "try:\n",
    "    for f in transfer['files']:\n",
    "        path = files[f['name']+':'+str(f['size'])]['path']\n",
    "        size = files[f['name']+':'+str(f['size'])]['size']\n",
    "        # putChunks\n",
    "        if debug:\n",
    "            print('putChunks: '+path)\n",
    "        with open(path, mode='rb', buffering=0) as fin:\n",
    "            for offset in range(0, size, upload_chunk_size):\n",
    "                if progress:\n",
    "                    print('Uploading: '+path+' '+str(offset)+'-'+str(min(offset +\n",
    "                          upload_chunk_size, size))+' '+str(round(offset/size*100))+'%')\n",
    "                data = fin.read(upload_chunk_size)\n",
    "                # print(data)\n",
    "                putChunk(transfer, f, data, offset)\n",
    "\n",
    "        # fileComplete\n",
    "        if debug:\n",
    "            print('fileComplete: '+path)\n",
    "        fileComplete(transfer, f)\n",
    "        if progress:\n",
    "            print('Uploading: '+path+' '+str(size)+' 100%')\n",
    "\n",
    "    # transferComplete\n",
    "    if debug:\n",
    "        print('transferComplete')\n",
    "    transferComplete(transfer)\n",
    "    if progress:\n",
    "        print('Upload Complete')\n",
    "\n",
    "except Exception as inst:\n",
    "    print(type(inst))\n",
    "    print(inst.args)\n",
    "    print(inst)\n",
    "\n",
    "    # deleteTransfer\n",
    "    if debug:\n",
    "        print('deleteTransfer')\n",
    "    deleteTransfer(transfer)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a9d143fc-a528-4886-9c70-8cbfdd8b4546",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "34feda0c-aa35-4863-a6a8-bd205de2decb",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e0179d24-3cd6-4994-9868-289295d2869b",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
